File: LeetCode/25.reverse-nodes-in-k-group.cpp
/*
 * @lc app=leetcode id=25 lang=cpp
 *
 * [25] Reverse Nodes in k-Group
 */
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class SolutionA
{
    // Seperate head-insertion method.
public:
    ListNode *reverseKGroup(ListNode *head, int k)
    {
        ListNode dummy(0);
        dummy.next = head;
        int count = 0;
        ListNode *prev = &dummy, *end = head;
        while (end != nullptr)
        {
            ++count;
            end = end->next;
            if (count == k)
            {
                ListNode *tmpHead = prev->next;
                ListNode *tmpEnd = tmpHead;
                ListNode *nextNode = tmpHead->next;
                tmpHead->next = end;
                while (nextNode != end)
                {
                    ListNode *tmp = nextNode->next;
                    nextNode->next = tmpHead;
                    tmpHead = nextNode;
                    nextNode = tmp;
                }
                prev->next = tmpHead;
                prev = tmpEnd;
                count = 0;
            }
        }
        return dummy.next;
    }
};

class Solution
{
    // Recursive method;
    ListNode *helper(ListNode *head, int k)
    {
        ListNode *end = head;
        int count = k;
        for (; end != nullptr && count != 0; end = end->next, --count);
        if (count == 0)
        {
            ListNode *nextNode = head->next;
            head->next = helper(end,k);
            while (nextNode != end)
            {
                ListNode *tmp = nextNode->next;
                nextNode->next = head;
                head = nextNode;
                nextNode = tmp;
            }
        }
        return head;
    }

public:
    ListNode *reverseKGroup(ListNode *head, int k)
    {
        return helper(head, k);
    }
};

File: lc-archived/java/25.reverse-nodes-in-k-group.java
/*
 * @lc app=leetcode id=25 lang=java
 *
 * [25] Reverse Nodes in k-Group
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        if(head == null||k<=1)
            return head;
        ListNode dummy = new ListNode(0, head);
        ListNode prevRange = dummy;
        while(prevRange.next != null){
            ListNode rangeEnd = prevRange;
            for(int i = 0;i<k && rangeEnd != null;++i){rangeEnd = rangeEnd.next;}
            if(rangeEnd == null)break;

            ListNode prev = prevRange.next;
            for(int i = 0;i<k-1;++i){
                ListNode ele = prev.next;
                prev.next = ele.next;
                ele.next=prevRange.next;
                prevRange.next=ele;
            }
            prevRange = prev;
        }
        return dummy.next;
    }

    private void printList(ListNode head){
        while(head != null){
            System.out.print(head.val+" ");
            head = head.next;
        }
        System.out.println();
    }
}

class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        if(head == null||k<=1)
            return head;

        ListNode cur = head;
        for(int i = 0;i<k-1 && cur != null;++i){cur = cur.next;}
        if(cur == null)return head;

        ListNode rest = cur.next;
        cur = head;
        head = reverseKGroup(rest, k);
        for(int i = 0;i<k;++i){
            ListNode tmp = cur.next;
            cur.next=head;
            head = cur;
            cur = tmp;
        }
        return head;

    }
}
// @lc code=end

